### PR TITLE
schema: add json defs for modules e-install (SC-651)

### PR DESCRIPTION
**Pull request by TheRealFalcon**
## Proposed Commit Message

```
schema: add json defs for modules e-install

* cc_emit_upstart (removal)
* cc_fan
* cc_final_message
* cc_foo
* cc_growpart
* cc_grub_dpkg
* cc_install_hotplug
```

## Additional Context
Removed 'cc_emit_upstart' entirely. According to https://upstart.ubuntu.com/ , the only place we'll see upstart anymore is trusty and RHEL 6, neither of which will ever see a new upstream change

### CODE REVIEW

File: cloudinit/config/cc_grub_dpkg.py [hunk 1]
```diff
@@ -121,14 +126,16 @@ def handle(name, cfg, _cloud, log, _args):
         return
 
     idevs = util.get_cfg_option_str(mycfg, "grub-pc/install_devices", None)
-    idevs_empty = util.get_cfg_option_str(
-        mycfg, "grub-pc/install_devices_empty", None
+    idevs_empty = util.get_cfg_option_bool(
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_grub_dpkg.py:129]: Small change here to accommodate deprecating the string option here. Also had to modify the test for this module.

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Looks pretty good. Inline comments throughout. I don't feel too strongly about the cc_growpart logic, but I think we we are truly going to deprecate some of these values or types, we should per logging deprecation warnings about them in this release.

File: cloudinit/config/cloud-init-schema.json [hunk 2]
```diff
@@ -545,6 +545,127 @@
...
+      "type": "object",
+      "properties": {
+        "growpart": {
+          "type": "object",
+          "additionalProperties": false,
+          "properties": {
+            "mode": {
+              "oneOf": [
+                {"type": "boolean"},
+                {"enum": ["auto", "growpart", "gpart", "off"]}
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cloudinit/config/cloud-init-schema.json:587]: Nice! Didn't know it was possible to have enum with multiple types inside

**Comment by TheRealFalcon** [comment]
Pushed next set from comments.

> I think we we are truly going to deprecate some of these values or types, we should per logging deprecation warnings about them in this release

I think I added warnings for everything here

File: cloudinit/config/cc_fan.py [hunk 3]
```diff
@@ -18,32 +22,35 @@
     - write ``config_path`` with the contents of the ``config`` key
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cloudinit/config/cc_fan.py]: Done

File: cloudinit/config/cc_final_message.py [hunk 4]
```diff
@@ -5,36 +5,47 @@
...
+from textwrap import dedent
+
+from cloudinit import templater, util, version
+from cloudinit.config.schema import MetaSchema, get_meta_doc
+from cloudinit.distros import ALL_DISTROS
+from cloudinit.settings import PER_ALWAYS
 
+MODULE_DESCRIPTION = """\
 This module configures the final message that cloud-init writes. The message is
 specified as a jinja template with the following variables set:
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cloudinit/config/cc_final_message.py:19]: Done

**Review by blackboxsw — APPROVED** [review]
Well done, thanks for the updates here.

**Pull request merged**